### PR TITLE
[distributed] Fix broadcast subgroup gradient rank comparison

### LINKED ISSUE #130515: [Bug Report] inconsistent global src and local rank in dist.nn.functional.broadcast
In torch.distributed.nn.functional._Broadcast, the below lines check whether the src and the rank are the same:

https://github.com/pytorch/pytorch/blob/dfd1d1971ea1265c597124b7e75fe5c8dd5a45b4/torch/distributed/nn/functional.py#L241-L242

However, the src here is global rank in the default process group WORLD (required by dist.broadcast), while the ctx.rank 

https://github.com/pytorch/pytorch/blob/dfd1d1971ea1265c597124b7e75fe5c8dd5a45b4/torch/distributed/nn/functional.py#L231

is a local rank under the provided process group by the function argument.

Thus this line fails to check whether the rank is the src, which results in all gradient to be zero for subgroups.

### PR DESCRIPTION
**Pull request by fhshaik**
## Summary

Fixes #130515.

`torch.distributed.nn.functional.broadcast` stores the current rank during the forward pass and uses it in backward to keep the reduced gradient only on the source rank. The source passed to `broadcast(src=...)` is a global rank, but the implementation stored a process-group-local rank. For subgroups whose local and global ranks differ, this comparison incorrectly zeroed the source gradient.

Store the global rank so both sides of the backward comparison use the same rank namespace. Add a two-process Gloo regression test using subgroup `[1]`, where source global rank 1 is subgroup-local rank 0.

## Test plan

- Two-process Gloo reproducer against unpatched PyTorch 2.13: fails because the source gradient is all zeros.
- The same reproducer with this patch: passes with the expected all-ones gradient.
- `python3 -m py_compile torch/distributed/nn/functional.py test/distributed/test_c10d_spawn_gloo.py`
- `git diff --check`

## AI assistance

This change was developed with assistance from OpenAI Codex. I reviewed the implementation, regression test, and diff and take responsibility for the contribution.

### CODE REVIEW

**Review by d4l3k — APPROVED** [review]
LGTM
